src/libraries/System.Private.CoreLib/src/System/IntPtr.cs

```diff
@@ -34,136 +33,125 @@ namespace System
+        public IntPtr(long value)
         {
 #if TARGET_64BIT
-            _value = (void*)value;
+            _value = (nint)value;
 #else
-            _value = (void*)checked((int)value);
+            _value = checked((nint)value);
 #endif
         }
 
         [CLSCompliant(false)]
         [NonVersionable]
         public unsafe IntPtr(void* value)
         {
-            _value = value;
+            _value = (nint)value;
         }
 
-        private unsafe IntPtr(SerializationInfo info, StreamingContext context)
+        private IntPtr(SerializationInfo info, StreamingContext context)
         {
-            long l = info.GetInt64("value");
+            long value = info.GetInt64("value");
 
-            if (Size == 4 && (l > int.MaxValue || l < int.MinValue))
+#if TARGET_32BIT
+            if ((value > int.MaxValue) || (value < int.MinValue))
+            {
                 throw new ArgumentException(SR.Serialization_InvalidPtrValue);
+            }
+#endif
 
-            _value = (void*)l;
+            _value = (nint)value;
         }
 
-        unsafe void ISerializable.GetObjectData(SerializationInfo info, StreamingContext context)
+        void ISerializable.GetObjectData(SerializationInfo info, StreamingContext context)
         {
             ArgumentNullException.ThrowIfNull(info);
 
-            info.AddValue("value", ToInt64());
+            long value = _value;
+            info.AddValue("value", value);
         }
 
-        public override unsafe bool Equals([NotNullWhen(true)] object? obj) =>
-            obj is IntPtr other &&
-            _value == other._value;
+        public override bool Equals([NotNullWhen(true)] object? obj) => (obj is nint other) && Equals(other);
 
-        public override unsafe int GetHashCode()
+        public override int GetHashCode()
         {
 #if TARGET_64BIT
-            long l = (long)_value;
-            return unchecked((int)l) ^ (int)(l >> 32);
+            long value = _value;
+            return value.GetHashCode();
 #else
-            return unchecked((int)_value);
+            return (int)_value;
 #endif
         }
 
         [NonVersionable]
-        public unsafe int ToInt32()
+        public int ToInt32()
         {
 #if TARGET_64BIT
-            long l = (long)_value;
-            return checked((int)l);
+            return checked((int)_value);
 #else
             return (int)_value;
 #endif
         }
 
         [NonVersionable]
-        public unsafe long ToInt64() =>
-            (nint)_value;
+        public long ToInt64() => _value;
 
         [NonVersionable]
-        public static unsafe explicit operator IntPtr(int value) =>
-            new IntPtr(value);
+        public static explicit operator nint(int value) => value;
 
         [NonVersionable]
-        public static unsafe explicit operator IntPtr(long value) =>
-            new IntPtr(value);
+        public static explicit operator nint(long value) => checked((nint)value);
```

Comment:
For my education, how doesn't this result in an infinite loop?

Reply:
The same as for `Int32`. Basically the C# compiler ignores all "user-defined operators" for the primitive types and uses its own "built-in" operators instead.

So the user-defined method basically only exists for "back-compat" and can only be invoked indirectly such as via reflection with the C# compiler just emitting `conv.ovf.i` for `checked((nint)value)`